profiles/killbill/src/main/java/org/killbill/billing/server/notifications/PushNotificationListener.java

```diff
@@ -112,13 +137,104 @@ public Response onCompleted(final Response response) throws Exception {
                         }
                     });
             response = futureStatus.get(timeoutSec, TimeUnit.SECONDS);
+        } catch (final TimeoutException toe) {
+            saveRetryPushNotificationInQueue(tenantId, url, body);
+            return false;
         } catch (final Exception e) {
             log.warn("Failed to push notification url='{}', tenantId='{}'", url, tenantId, e);
             return false;
         }
+
+        if ((response.getStatusCode() == 301) || (response.getStatusCode() == 500)) {
```

Comment:
@pierre @sbrossie I refactor the code as Pierre suggested. Can you please confirm that you want to schedule a retry for all exceptions? Code will look like this:
```Java
try {
    final ListenableFuture<Response> futureStatus =
        builder.execute(new AsyncCompletionHandler<Response>() {
            @Override
            public Response onCompleted(final Response response) throws Exception {
                return response;
        }
    });
    response = futureStatus.get(timeoutSec, TimeUnit.SECONDS);
} catch (final Exception e) {
    log.warn("Failed to push notification url='{}', tenantId='{}'", url, tenantId, e);
    saveRetryPushNotificationInQueue(tenantId, url, notification, attemptRetryNumber);
    return false;
}
```